Alignment/include/ActsAlignment/Kernel/detail/AlignmentEngine.hpp

```diff
@@ -0,0 +1,302 @@
+  measurementStates.reserve(15);
+  // Number of smoothed states on the track
+  size_t nSmoothedStates = 0;
+  // Number of alignable surfaces on the track
+  size_t nAlignSurfaces = 0;
+
+  // Visit the track states on the track
+  multiTraj.visitBackwards(entryIndex, [&](const auto& ts) {
+    // Remember the number of smoothed states
+    if (ts.hasSmoothed()) {
+      nSmoothedStates++;
+    } else {
+      // @FIXME: is this possible?
+      return true;
+    }
+
+    // Only measurement states matter (we can't align non-measurement states,
+    // no?)
+    if (not ts.typeFlags().test(TrackStateFlag::MeasurementFlag)) {
+      return true;
+    }
+    // Check if the reference surface is to be aligned
+    // @Todo: consider the case when some of the Dofs are fixed for one surface
+    bool isAlignable = false;
+    const auto surface = &ts.referenceSurface();
+    auto it = idxedAlignSurfaces.find(surface);
+    if (it != idxedAlignSurfaces.end()) {
+      isAlignable = true;
+      // Remember the surface and its index
+      alignState.alignedSurfaces[surface].first = it->second;
+      nAlignSurfaces++;
+    }
+    // Rember the index of the state within the trajectory and whether it's
+    // alignable
+    measurementStates.push_back({ts.index(), isAlignable});
+    // Add up measurement dimension
+    alignState.measurementDim += ts.calibratedSize();
+    return true;
+  });
+
+  // Return now if the track contains no alignable surfaces
+  if (nAlignSurfaces == 0) {
+    return alignState;
+  }
+
+  // The alignment degree of freedom
+  alignState.alignmentDof = eAlignmentSize * nAlignSurfaces;
+  // Dimension of global track parameters (from only measurement states)
+  alignState.trackParametersDim = eBoundSize * measurementStates.size();
+
+  // Initialize the alignment matrixs with components from the measurement
+  // states
+  // The measurement covariance
+  alignState.measurementCovariance = ActsDynamicMatrix::Zero(
+      alignState.measurementDim, alignState.measurementDim);
+  // The bound parameters to measurement projection matrix
+  alignState.projectionMatrix = ActsDynamicMatrix::Zero(
+      alignState.measurementDim, alignState.trackParametersDim);
+  // The derivative of residual w.r.t. alignment parameters
+  alignState.alignmentToResidualDerivative = ActsDynamicMatrix::Zero(
+      alignState.measurementDim, alignState.alignmentDof);
+  // The track parameters covariance
+  alignState.trackParametersCovariance = ActsDynamicMatrix::Zero(
+      alignState.trackParametersDim, alignState.trackParametersDim);
+  // The residual
+  alignState.residual = ActsDynamicVector::Zero(alignState.measurementDim);
+
+  // The dimension of provided global track parameters covariance should be same
+  // as eBoundSize * nSmoothedStates
+  assert(globalTrackParamsCov.first.rows() ==
+             globalTrackParamsCov.first.cols() and
+         globalTrackParamsCov.first.rows() == eBoundSize * nSmoothedStates);
```

Comment:
Is this meant to fail only in debug mode ?

Reply:
Acutually, I don't expect it to fail. So removed now.

Reply:
Good.